### PR TITLE
add algorithm edit list

### PR DESCRIPTION
**Pull request by ticoann**
related to #6720

Just copied the list of algorithm and editable field from reqmgr1.

### CODE REVIEW

**Comment by ticoann** [comment]
@amaltaro, Alan, please review. Especially please check the algorithm list and editable splitting parameters whether is missing parameters or missing algorithm. Thanks,

File: src/python/WMCore/ReqMgr/Service/RequestAdditionalInfo.py [hunk 1]
```diff
@@ -6,11 +6,51 @@
...
 import WMCore.ReqMgr.Service.RegExp as rx
-from WMCore.REST.Format import JSONFormat
+from WMCore.REST.Format import JSONFormat, PrettyJSONFormat
 
 from WMCore.Wrappers import JsonWrapper
 from WMCore.WMSpec.WMWorkload import WMWorkloadHelper
 
+#TODO: need to get algorithm proper way
+ALGO_DICT = {"FileBased": {"files_per_job": 0,
+                           "include_parents": False},
```
Comment (amaltaro) [inline src/python/WMCore/ReqMgr/Service/RequestAdditionalInfo.py]: I tend to say "include_parents" should not be a splitting changeable parameter. It's defined during creation time with "IncludeParents" and that's it.

File: src/python/WMCore/ReqMgr/Service/RequestAdditionalInfo.py [hunk 2]
```diff
@@ -6,11 +6,51 @@
...
+             "LumiBased":{"lumis_per_job": 0, 
+                          "halt_job_on_file_boundaries": True,
+                           "include_parents": False},
+             "EventAwareLumiBased": {"avg_events_per_job": 0,
+                                     "max_events_per_lumi": 0,
+                                     "halt_job_on_file_boundaries": True,
+                                     "include_parents": False},
+             "EventBased": {"events_per_job": 0,
+                            "events_per_lumi": 0,
+                            "lheInputFiles": 0,
```
Comment (amaltaro) [inline src/python/WMCore/ReqMgr/Service/RequestAdditionalInfo.py]: this lhe is boolean

File: src/python/WMCore/ReqMgr/Service/RequestAdditionalInfo.py [hunk 3]
```diff
@@ -208,4 +248,45 @@ def post(self, name):
...
+    def __init__(self, app, api, config, mount):
+        # main CouchDB database where requests/workloads are stored
+        RESTEntity.__init__(self, app, api, config, mount)
+        self.reqdb_url = "%s/%s" % (config.couch_host, config.couch_reqmgr_db)
+    
+    
+    def _validate_args(self, param, safe):
+        # TODO: need proper validation but for now pass everything
+        args_length = len(param.args)
+        if args_length > 0 :
```
Comment (amaltaro) [inline src/python/WMCore/ReqMgr/Service/RequestAdditionalInfo.py]: Isn't it `if args_length < 1:`?
BTW, for my education, besides handling different validations (which is not the case for this class), what's the motivation not to put the actual validation code into the validate() method?

File: src/python/WMCore/ReqMgr/Service/RequestAdditionalInfo.py [hunk 4]
```diff
@@ -6,11 +6,51 @@
...
+    fdict = {"taskType": task_type}
+    param_list = []
+    for algo in algo_list:
+        param = {"alogrithm": algo}
+        param.update(ALGO_DICT[algo])
+        param_list.append(param)
+    fdict["splitParamList"] = param_list
+    return fdict
+        
+SPILTT_ALGOS = [format_algo_list("Processing", PROCESSING_ALGO),
```
Comment (amaltaro) [inline src/python/WMCore/ReqMgr/Service/RequestAdditionalInfo.py]: Should read `SPLIT_ALGOS`, the same in the class get method

File: src/python/WMCore/ReqMgr/Service/RequestAdditionalInfo.py [hunk 5]
```diff
@@ -6,11 +6,51 @@
...
+                format_algo_list("Procuction", PRODUCTION_ALGO),
+                format_algo_list("Skim", SKIM_ALGO),
+                format_algo_list("Harviest", HARVEST_ALGO),
```
Comment (amaltaro) [inline src/python/WMCore/ReqMgr/Service/RequestAdditionalInfo.py]: Harvest

File: src/python/WMCore/ReqMgr/Service/RequestAdditionalInfo.py [hunk 6]
```diff
@@ -6,11 +6,51 @@
...
+    param_list = []
+    for algo in algo_list:
+        param = {"alogrithm": algo}
+        param.update(ALGO_DICT[algo])
+        param_list.append(param)
+    fdict["splitParamList"] = param_list
+    return fdict
+        
+SPILTT_ALGOS = [format_algo_list("Processing", PROCESSING_ALGO),
+                format_algo_list("Procuction", PRODUCTION_ALGO),
```
Comment (amaltaro) [inline src/python/WMCore/ReqMgr/Service/RequestAdditionalInfo.py]: Production

File: src/python/WMCore/ReqMgr/Service/RequestAdditionalInfo.py [hunk 7]
```diff
@@ -6,11 +6,51 @@
...
+PROCESSING_ALGO = ["LumiBased", "EventAwareLumiBased", "EventBased", "FileBased"]
+PRODUCTION_ALGO = ["EventBased"]
+SKIM_ALGO = ["FileBased", "TwoFileFileBased"]
+HARVEST_ALGO = ["Harvest"]
+
+def format_algo_list(task_type, algo_list):
+    fdict = {"taskType": task_type}
+    param_list = []
+    for algo in algo_list:
+        param = {"alogrithm": algo}
```
Comment (amaltaro) [inline src/python/WMCore/ReqMgr/Service/RequestAdditionalInfo.py]: algorithm

**Comment by amaltaro** [comment]
Besides the few comments, it sounds good. It would return something like:
```
{'splitParamList': [{'files_per_job': 0, 'include_parents': False, 'alogrithm': 'FileBased'}, {'two_files_per_job': 0, 'include_parents': False, 'alogrithm': 'TwoFileFileBased'}], 'taskType': 'Skim'}
```
for Skim

File: src/python/WMCore/ReqMgr/Service/RequestAdditionalInfo.py (hunk 3 above)
Comment (ticoann) [inline src/python/WMCore/ReqMgr/Service/RequestAdditionalInfo.py]: It is raising exception if args is passed so basically it says it shouldn't pass the args.
The reason  it separates _validate_args from validation function is in general, there are different validation requires for different verbs (GET, POST etc) so I would like to separate that. However in this case we have only GET so we could move the code in validation. But I am thinking to add PUT method so we can add new Algorithm on the fly if necessary.  (Maybe never happen. :smile:)

**Comment by ticoann** [comment]
@amaltaro Thank you very much Alan, I fixed things you pointed out. I shouldn't create the patch at night. :smile: Please review again. We have a lot more algorithms exist in WMCore. So the others are not used?
https://github.com/dmwm/WMCore/tree/master/src/python/WMCore/JobSplitting

**Comment by ticoann** [comment]
@amaltaro, After talking to Valentin, I changed the API to give all filtered result for splitting params for web editing. Please review.

https://reqmgr2-dev.cern.ch/reqmgr2/data/splitting/web_form/sryu_TaskChain_Task_Multicore_Sryu_reqmgr2_160318_050229_8964

To get all the params.
https://reqmgr2-dev.cern.ch/reqmgr2/data/splitting/sryu_TaskChain_Task_Multicore_Sryu_reqmgr2_160318_050229_8964

File: src/python/WMCore/ReqMgr/Service/RequestAdditionalInfo.py [hunk 8]
```diff
@@ -6,11 +6,72 @@
...
+             "LumiBased": {"lumis_per_job": '', 
+                          "halt_job_on_file_boundaries": True},
+             "EventAwareLumiBased": {"avg_events_per_job": '',
+                                     "max_events_per_lumi": '',
+                                     "halt_job_on_file_boundaries": True},
+             "EventBased": {"events_per_job": '',
+                            "events_per_lumi": '',
+                            "lheInputFiles": False},
+             "Harvest": {"periodic_harvest_interval": ''},
+             "ParentlessMergeBySize": {"min_merge_size": '',
```
Comment (amaltaro) [inline src/python/WMCore/ReqMgr/Service/RequestAdditionalInfo.py]: We shouldn't touch Merge algos. AFAIK WMCore enforces it according to the task type and job spliting for the Processing/Production task

**Comment by amaltaro** [comment]
The special handling for default_algo (the initial algo for that workflow) is to show the original job splitting values, isn't it?
The Merge splitting we _usually_ do not touch, I'm not sure we should show them in the Gui... and AFAIK we use mostly ParentlessMergeBySize and WMBSMergeBySize.

For the other algos, if they are not used by T0, then they are legacy: FixedDelay, MergeBySize (not sure), MinFileBased, RunBased (hmm, sounds interesting), SiblingProcessingBased (what?), SizeBased, SplitFileBased, TwoFileAndEventBased ...oh well, where these ideas come from...

File: src/python/WMCore/ReqMgr/Service/RequestAdditionalInfo.py (hunk 8 above)
Comment (amaltaro) [inline src/python/WMCore/ReqMgr/Service/RequestAdditionalInfo.py]: Nevermind this comment, I'm deleting it

**Comment by amaltaro** [comment]
Ok, looking at your web_form:
https://reqmgr2-dev.cern.ch/reqmgr2/data/splitting/web_form/sryu_TaskChain_Task_Multicore_Sryu_reqmgr2_160318_050229_8964

I changed my mind, there may be use cases where we want to change the Merge splitting parameters, but I still think the algo itself should be unchangeable by the user.

About Celanup and LogCollect, I'd keep the web form clean and simply not show them. This would be one complain that I have about the reqmgr splitting page, it has way too much info...

**Comment by ticoann** [comment]
Thanks, Alan, I changed the api as you suggested.
1. remove cleanup and log collect update
2. For merge only parameter changes but not algorithm.

https://reqmgr2-dev.cern.ch/reqmgr2/data/splitting/web_form/sryu_TaskChain_Task_Multicore_Sryu_reqmgr2_160318_050229_8964

Please review.

**Comment by ticoann** [comment]
Hi Alan, I refactored code and updated.

Looking at the test TaskChain, I see something weird.

https://reqmgr2-dev.cern.ch/reqmgr2/data/splitting/sryu_McMTaskChain2_Sryu_reqmgr2_160318_050204_3263

There is on Merge is set with "EventAwareLumiBased",
Is that a mistake?
Also there empty algorithm, I guess in that case it takes default algorithm
```
{
  "splitAlgo": "EventAwareLumiBased", 
  "splitParams": {
    "algorithm": "EventAwareLumiBased", 
    "initial_lfn_counter": 0, 
    "max_events_per_lumi": 20000, 
    "trustSitelists": false, 
    "type": "Merge"
  }, 
  "taskName": "/sryu_McMTaskChain2_Sryu_reqmgr2_160318_050204_3263/EXO-RunIIFall15DR76-03799_0/EXO-RunIIFall15DR76-03799_1/EXO-RunIIFall15DR76-03799_1MergeDQMoutput", 
  "taskType": "Merge"
}
```

**Comment by amaltaro** [comment]
There is something wrong here, both first task and a merge task have an empty list of algos (besides a Merge task showing up a Processing algo).

I injected a good test case:
https://reqmgr2-dev.cern.ch/reqmgr2/data/splitting/amaltaro_TaskChain_MC_DMWM_TEST_SplitAPI_160324_113023_1915
Task1: should default to EventBased, 300 evts per Job
Task2: should default to EventAwareLumiBased, 600 evts per job (6l)
Task3: should be LumiBased, as enforced in the schema, 1000 evts per job (10l)

**Comment by ticoann** [comment]
Yes something is definitely wrong.
From schema, The SplitAlgo is missing for Task1 and Task2

https://reqmgr2-dev.cern.ch/couchdb/_utils/document.html?reqmgr_workload_cache/amaltaro_TaskChain_MC_DMWM_TEST_SplitAPI_160324_113023_1915

**Comment by amaltaro** [comment]
BTW @vkuznet I connected to reqmgr2-dev and I like it much more now, it's more compact and seems cleaner ;)

**Comment by ticoann** [comment]
@amaltaro, Alan, could you point me the json file you used to create
amaltaro_TaskChain_MC_DMWM_TEST_SplitAPI_160324_113023_1915?

Is this one?
https://github.com/dmwm/WMCore/blob/master/test/data/ReqMgr/requests/DMWM/TaskChain_MC.json

Or something else. It seems above one has algorithm specified.

Anyway, it seems the problem is not in this patch but (it might be the way we are setting default)
I updated the code so instead of allowing empty algorithm, we will through error.

Tested with above json and seems to work fine. I am not sure how I created my workflow.

**Comment by amaltaro** [comment]
It's basically the same template, but with a few tweaks. You can see it here:
http://amaltaro.web.cern.ch/amaltaro/forSeangChan/TaskChain_MC.json

BTW, the way we set default works just fine, it must be something related to reqmgr2 only.

**Comment by ticoann** [comment]
Hi Alan,
I created two workflows  with your json.
reqmgr1 - sryu_TaskChain_MC_alan_Sryu_reqmgr1_160324_185923_7359
reqmgr2 - sryu_TaskChain_MC_alan_Sryu_reqmgr1_160324_185907_1165

Both seems to be same and fine. Not sure could you try to create again.

https://reqmgr2-dev.cern.ch/reqmgr2/data/splitting/sryu_TaskChain_MC_alan_Sryu_reqmgr1_160324_185923_7359

https://reqmgr2-dev.cern.ch/reqmgr2/data/splitting/sryu_TaskChain_MC_alan_Sryu_reqmgr1_160324_185907_1165

**Comment by ticoann** [comment]
Merging it now. Will further debug if the problem occurs

**Pull request merged**